block/manager.go

```diff
@@ -195,50 +218,75 @@ func (m *Manager) SyncLoop(ctx context.Context) {
 	}
 }
 
+// RetrieveLoop is responsible for interacting with DA layer.
 func (m *Manager) RetrieveLoop(ctx context.Context) {
+	// waitCh is used to signal the retrieve loop, that it should process next blocks
+	// retrieveCond can be signalled in completely async manner, and goroutine below
+	// works as some kind of "buffer" for those signals
+	waitCh := make(chan interface{})
+	go func() {
+		for {
+			m.retrieveMtx.Lock()
+			m.retrieveCond.Wait()
+			waitCh <- nil
+			m.retrieveMtx.Unlock()
+			if ctx.Err() != nil {
+				return
+			}
+		}
+	}()
+
 	for {
 		select {
-		case <-m.retrieveCh:
-			target := atomic.LoadUint64(&m.syncTarget)
-			for h := m.store.Height() + 1; h <= target; h++ {
-				m.logger.Debug("trying to retrieve block from DALC", "height", h)
-				m.mustRetrieveBlock(ctx, h)
+		case <-waitCh:
+			for {
+				daHeight := atomic.LoadUint64(&m.daHeight)
+				m.logger.Debug("retrieve", "daHeight", daHeight)
+				err := m.processNextDABlock()
+				if err != nil {
+					m.logger.Error("failed to retrieve block from DALC", "daHeight", daHeight, "errors", err.Error())
+					break
+				}
+				atomic.AddUint64(&m.daHeight, 1)
```

Comment:
Wait, are you sure this is the right use of atomics?

What if `&m.daHeight` changes in between loading it and adding `1` to it here? Then you'd be adding `2` to it.

Reply:
This is the only place where `m.daHeight` is changed. It's atomic change, because we're reading this variable in several other places (in other goroutines). TBH the `LoadUint64` in the beginning of the loop is not necessary (normal read should be fine), but I decided to handle this variable "atomically" everywhere for consistency.

Reply:
My concern is if e.g. `&m.daHeight` is `N`, then it could be read as `N` in two separate threads and incremented twice to `N+2`, completely bypassing `N+1`. Is that ever possible, if is so, is it a problem?

Reply:
It's not possible right now, as there is only one thread incrementing this variable, and I don't see a reason to add more threads/goroutines.